build: Drop option to disable hardening.

davidgumberg opened the pull request:
Follow up to #32038 which dropped `NO_HARDEN` from depends builds, this PR drops the `ENABLE_HARDENING` build option since disabling hardening of binaries should not be a supported or maintained use case. With this change, hardening flags are always enabled.

Individual hardening flags and options can still be disabled by appending flags, e.g.:

```bash
cmake -B build \
  -DAPPEND_CPPFLAGS='-U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=0 -fno-stack-protector -fcf-protection=none -fno-stack-clash-protection' \
  -DAPPEND_LDFLAGS='-Wl,-z,lazy -Wl,-z,norelro -Wl,-z,noseparate-code'
```

There is an issue with NetBSD 10.0's dynamic linker that makes one of the hardening linker flags, `-z separate-code`, [problematic](https://github.com/bitcoin/bitcoin/pull/28724#issuecomment-2589347934), so this PR also introduces a check to prevent the use of this flag in NetBSD versions < 11.0, (where this issue is [fixed](https://github.com/NetBSD/src/commit/acf7fb3abf88f29f6b15c894b6b18cadc12763c0)). The fix for this [might be backported](https://mail-index.netbsd.org/tech-userlevel/2023/01/05/msg013670.html) to NetBSD 10.0.

I suggest reviewing the diff with whitespace changes hidden (`git diff -w` or using github's hide whitespace option)

davidgumberg committed:
build: Drop option for disabling hardening

Building unhardened executables is not a supported use case that should
be maintained and those that want unhardened executables can still
override them by appending disable flags.

For example:

cmake -B build -DAPPEND_CPPFLAGS='-U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=0 -fno-stack-protector -fcf-protection=none -fno-stack-clash-protection' -DAPPEND_LDFLAGS='-Wl,-z,lazy -Wl,-z,norelro -Wl,-z,noseparate-code'

maflcko reviewed (commented):
review ACK ecf2046d4b5c43ddf64f62f09cd3ed70dd5caafb 📮

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: review ACK ecf2046d4b5c43ddf64f62f09cd3ed70dd5caafb 📮
kuLcmyw1Kv99uptN9UVbNG0JwSyCRffWoUy3VFb9NZGP/v7+SpCYNa5PrWtPOGQHROhCkU6Rgf2BpXdp95KKAg==
```

</details>

laanwj reviewed (approved):
Concept and code review ACK ecf2046d4b5c43ddf64f62f09cd3ed70dd5caafb
rationale: https://github.com/bitcoin/bitcoin/pull/32038#issuecomment-2717727013 https://github.com/bitcoin/bitcoin/pull/32038#issuecomment-2720530698

vasild commented on CMakeLists.txt:
```diff
@@ -481,63 +480,61 @@ try_append_cxx_flags("-fmacro-prefix-map=A=B" TARGET core_interface SKIP_LINK
-    try_append_linker_flag("/HIGHENTROPYVA" TARGET hardening_interface)
-    try_append_linker_flag("/NXCOMPAT" TARGET hardening_interface)
-  else()
+add_library(hardening_interface INTERFACE)
+target_link_libraries(core_interface INTERFACE hardening_interface)
```
`hardening_interface` is not needed anymore? The surrounding code appends stuff directly to `core_interface`. Maybe remove these 2 lines and `s/hardening_interface/core_interface/` all over the place?

davidgumberg replied:
Good catch, will address in a refactor follow-up, or if I touch here again.

davidgumberg replied:
Fixed in latest push

vasild reviewed (approved):
ACK ecf2046d4b5c43ddf64f62f09cd3ed70dd5caafb

hebasto reviewed (commented):
Approach ACK ecf2046d4b5c43ddf64f62f09cd3ed70dd5caafb.

The only case I'm aware of where some hardening flags cause an issue is on NetBSD 10, and this is expected to be fixed in NetBSD 11.

For more details, see:
- https://github.com/bitcoin/bitcoin/pull/28724#issuecomment-2585683053
- https://github.com/hebasto/bitcoin-core-nightly/pull/30

laanwj commented:
> The only case I'm aware of where some hardening flags cause an issue is on NetBSD 10, and this is expected to be fixed in NetBSD 11.

Right. Not sure if it's worth it in this specific case, as it's going to be fixed upstream soon, but imo in general if some flags (hardening or otherwise) are a problem on some platform this should be taken into account by the build system automatically instead of having the user disable hardening entirely, which is a hassle and creates a bigger risk than necessary.

davidgumberg commented:
> > The only case I'm aware of where some hardening flags cause an issue is on NetBSD 10, and this is expected to be fixed in NetBSD 11.
> 
> in general if some flags (hardening or otherwise) are a problem on some platform this should be taken into account by the build system automatically instead of having the user disable hardening entirely.

Okay, I think this branch will break building on NetBSD, and shouldn't be merged as-is, I'll follow up.

fanquake commented:
>  I'll follow up.

Do you want to split out 7d34c19853e7a5528d69c5f30580e7e9712e61f0? That can be merged now.

davidgumberg commented:
> Do you want to split out [7d34c19](https://github.com/bitcoin/bitcoin/commit/7d34c19853e7a5528d69c5f30580e7e9712e61f0)? That can be merged now.

Opened https://github.com/bitcoin/bitcoin/pull/32087

hebasto commented:
> > > The only case I'm aware of where some hardening flags cause an issue is on NetBSD 10, and this is expected to be fixed in NetBSD 11.
> > 
> > 
> > in general if some flags (hardening or otherwise) are a problem on some platform this should be taken into account by the build system automatically instead of having the user disable hardening entirely.
> 
> Okay, I think this branch will break building on NetBSD, and shouldn't be merged as-is, I'll follow up.

Perhaps I should clarify that only the `libbitcoinkernel.so` shared library is broken. All executables run fine, except that invoking `ldd` (or any other operation involving the dynamic linker) fails.

Therefore, I'm fine with dropping `NO_HARDEN`.

fanquake commented:
@davidgumberg want to rebase here?

davidgumberg committed:
build: refactor: hardening flags -> core_interface

davidgumberg committed:
build: Use `-z noseparate-code` on NetBSD < 11.0

This can be dropped once Bitcoin Core no longer supports NetBSD 10.0 or
if upstream fix is backported.

NetBSD's dynamic linker ld.elf_so < 11.0 supports exactly 2 `PT_LOAD`
segments and binaries linked with `-z separate-code` have 4 `PT_LOAD`
segments.

https://github.com/bitcoin/bitcoin/pull/28724#issuecomment-2589347934
https://mail-index.netbsd.org/tech-userlevel/2023/01/05/msg013666.html

davidgumberg commented:
Rebased to address @vasild feedback to drop `hardening_interface`, and added a check to `CMakeLists.txt` to avoid `-z separate-code` on NetBSD < 11.0 (thanks @hebasto). I've also updated the PR description.

hebasto commented on CMakeLists.txt:
```diff
@@ -481,63 +480,71 @@ try_append_cxx_flags("-fmacro-prefix-map=A=B" TARGET core_interface SKIP_LINK
+  # - https://github.com/bitcoin/bitcoin/pull/28724#issuecomment-2589347934
+  # - https://mail-index.netbsd.org/tech-userlevel/2023/01/05/msg013666.html
+  if(CMAKE_SYSTEM_NAME STREQUAL "NetBSD" AND CMAKE_SYSTEM_VERSION VERSION_LESS 11.0)
+      try_append_linker_flag("-Wl,-z,noseparate-code" TARGET core_interface)
```
style nit: two-spaces indent.

davidgumberg replied:
Thanks, fixed

hebasto reviewed (approved):
ACK 0e98ec819b65cfc1728a6aecf5e43a7c73756a66, I have reviewed the code and it looks OK.

Here are NetBSD CI jobs for this branch: https://github.com/hebasto/bitcoin-core-nightly/actions/runs/14036068740.

hebasto reviewed (approved):
re-ACK 77e553ab6a0a98d065884a83490f28eec6df0e23.

laanwj reviewed (approved):
re-ACK 77e553ab6a0a98d065884a83490f28eec6df0e23

fanquake commented:
cc @theuni

janb84 reviewed (commented):
ACK [77e553a](https://github.com/bitcoin/bitcoin/commit/77e553ab6a0a98d065884a83490f28eec6df0e23)

- code reviewed ✅
- build ✅
- tested ✅

fanquake commented on CMakeLists.txt:
```diff
@@ -481,63 +480,71 @@ try_append_cxx_flags("-fmacro-prefix-map=A=B" TARGET core_interface SKIP_LINK
+  # `PT_LOAD` segments and binaries linked with `-z separate-code`
+  # have 4 `PT_LOAD` segments.
+  # Relevant discussions:
+  # - https://github.com/bitcoin/bitcoin/pull/28724#issuecomment-2589347934
```
This whole block seems pretty verbose (we don't need to repeat all the details from the NetBSD mailinglist), and its not actually clear these workarounds are needed for anything other than for the experimental kernel library, according to https://github.com/bitcoin/bitcoin/pull/32071#issuecomment-2732279914? The linked comment in our repo just repeats what's in the mailing list post, so could also be dropped.

hebasto replied:
From the mentioned [#32071 (comment)](https://github.com/bitcoin/bitcoin/pull/32071#issuecomment-2732279914):

> Perhaps I should clarify that only the `libbitcoinkernel.so` shared library is broken. All executables run fine, except that invoking `ldd` (or any other operation involving the dynamic linker) fails.

I believe it's helpful to ensure that running `ldd bitcoind` produces no errors.

fanquake replied:
> I believe it's helpful to ensure that running ldd bitcoind produces no errors.

It's not really clear that disabling hardening based on (a buggy?) `ldd` is the correct tradeoff.

I also don't completely understand your comment in regards to the binaries running fine, but simultaneously, any operation using the dynamic linker (i.e running the binaries) failing.

davidgumberg replied:
> This whole block seems pretty verbose (we don't need to repeat all the details from the NetBSD mailinglist), and its not actually clear these workarounds are needed for anything other than for the experimental kernel library, according to https://github.com/bitcoin/bitcoin/pull/32071#issuecomment-2732279914? The linked comment in our repo just repeats what's in the mailing list post, so could also be dropped.

Agreed, but I am not sure if this is blocking and would prefer to only change this if retouching to avoid invalidating ACK's.

> It's not really clear that disabling hardening based on (a buggy?) `ldd` is the correct tradeoff.

That is fair, but only `separate-code` is being disabled, and while having `separate-code` enabled is strictly better in theory, if I understand correctly, in practice it is dubious whether or not there are any likely situations where an attacker will not be able to find a gadget in a text segment but find one in the page overlap between text and data segments.

Upstream discussion in LLD where `-noseparate-code` was made default (https://reviews.llvm.org/D64903):

> I don't think -z separate-code is a very effective defense against a ROP/JOP. If a program is not very small, it shouldn't be too hard to find a gadget from a text segment, and I don't think an additional piece of data that is at the remaining part of the last page of a text segment can significantly increase a risk. If your program is vulnerable to ROP/JOP, and gadgets are at predictable places, it's a game end regardless of an existence of some extra data at the end of a text segment.

luke-jr commented:
>this PR also introduces a check to prevent the use of this flag in NetBSD versions < 11.0, (where this issue is [fixed](https://github.com/NetBSD/src/commit/acf7fb3abf88f29f6b15c894b6b18cadc12763c0)). The fix for this [might be backported](https://mail-index.netbsd.org/tech-userlevel/2023/01/05/msg013670.html) to NetBSD 10.0.

If the fix is backported, presumably your change will still disable it?

vasild reviewed (approved):
ACK 77e553ab6a0a98d065884a83490f28eec6df0e23

sipa reviewed (commented):
Concept ACK

davidgumberg commented:
> > this PR also introduces a check to prevent the use of this flag in NetBSD versions < 11.0, (where this issue is [fixed](https://github.com/NetBSD/src/commit/acf7fb3abf88f29f6b15c894b6b18cadc12763c0)). The fix for this [might be backported](https://mail-index.netbsd.org/tech-userlevel/2023/01/05/msg013670.html) to NetBSD 10.0.
> 
> If the fix is backported, presumably your change will still disable it?

Yes, but I think that's okay, since only one flag of dubious hardening benefit is disabled on NetBSD. More discussion in this comment: https://github.com/bitcoin/bitcoin/pull/32071#discussion_r2040365957

musaHaruna reviewed (commented):
tested ACK [77e553](https://github.com/bitcoin/bitcoin/commit/77e553ab6a0a98d065884a83490f28eec6df0e23) 
Code reviewed and built successfully

fanquake commented:
Guix Build:
```bash
db95c26122204cf686ae4d0a7d6f2e2f2cd30d7ea0dce425f28bf4fa1af9fa12  guix-build-77e553ab6a0a/output/aarch64-linux-gnu/SHA256SUMS.part
3b5b3fd4b6ca8539e2ea5e50afdea37593f1f2e0c0b26851c20fc0505dd79a42  guix-build-77e553ab6a0a/output/aarch64-linux-gnu/bitcoin-77e553ab6a0a-aarch64-linux-gnu-debug.tar.gz
4c6b3a3949cc70e3f1a82fa38ed5c204c54074a79b267d67c02439279099c2ff  guix-build-77e553ab6a0a/output/aarch64-linux-gnu/bitcoin-77e553ab6a0a-aarch64-linux-gnu.tar.gz
b9d2d30e3b5df7ad303c951176feb18a18c42e43c95a038128ca10669bb4a6a9  guix-build-77e553ab6a0a/output/arm-linux-gnueabihf/SHA256SUMS.part
58cb80d16437acba0bbaa5df0a69521270dbc61b84a5b8284c18b4e96263b0a6  guix-build-77e553ab6a0a/output/arm-linux-gnueabihf/bitcoin-77e553ab6a0a-arm-linux-gnueabihf-debug.tar.gz
aa55ed2d8be3cfa760d373ab971ebf686c040785b0c64fd2e77779475a866165  guix-build-77e553ab6a0a/output/arm-linux-gnueabihf/bitcoin-77e553ab6a0a-arm-linux-gnueabihf.tar.gz
5b217641817eda09edbf4efdc74b9c110defcb4bdd85c9d37c91b0721dcf7674  guix-build-77e553ab6a0a/output/arm64-apple-darwin/SHA256SUMS.part
6453fb94f8c9015c080eda0d0eb3818b9c07e1073b7dae05758b7a0f9259aa15  guix-build-77e553ab6a0a/output/arm64-apple-darwin/bitcoin-77e553ab6a0a-arm64-apple-darwin-codesigning.tar.gz
ad867c154209ddb30b245f813a531406e55b0803ad100187f66ec22d8a8a8780  guix-build-77e553ab6a0a/output/arm64-apple-darwin/bitcoin-77e553ab6a0a-arm64-apple-darwin-unsigned.tar.gz
2b176309d5e44bc7c8cc8d6a352d47c59699ce449e566653fefc661c1e72ecda  guix-build-77e553ab6a0a/output/arm64-apple-darwin/bitcoin-77e553ab6a0a-arm64-apple-darwin-unsigned.zip
b433b9c2bd8d9258414edbd98f247771c2edcfa532eba8dc8bde10d93c98048b  guix-build-77e553ab6a0a/output/dist-archive/bitcoin-77e553ab6a0a.tar.gz
356a8f1c01fd3ea760ed6e120814f3fe9b9a4ce69b7c22b504369a702d776ef3  guix-build-77e553ab6a0a/output/powerpc64-linux-gnu/SHA256SUMS.part
fc42a3468634083cceaa0a606953e6199b0f0071ced469395bce98c763aee2a2  guix-build-77e553ab6a0a/output/powerpc64-linux-gnu/bitcoin-77e553ab6a0a-powerpc64-linux-gnu-debug.tar.gz
f52e9149e7e792592c48d31f1b0ce319c5ee887077157df2c6a6665fbebc8112  guix-build-77e553ab6a0a/output/powerpc64-linux-gnu/bitcoin-77e553ab6a0a-powerpc64-linux-gnu.tar.gz
39aefacf7be73c6dbf444532c3bc344ecebf9290cb6a80e1072bfb2e8eae6f0c  guix-build-77e553ab6a0a/output/riscv64-linux-gnu/SHA256SUMS.part
1c3ea57370a8e04dea9d86aac34af7e5ba118f2835431137e4475efeed5543c5  guix-build-77e553ab6a0a/output/riscv64-linux-gnu/bitcoin-77e553ab6a0a-riscv64-linux-gnu-debug.tar.gz
fb63f9bd863e5305ae0b385a2c2b1d0dbb61ef872d3fda1d8ca12fed9e23455b  guix-build-77e553ab6a0a/output/riscv64-linux-gnu/bitcoin-77e553ab6a0a-riscv64-linux-gnu.tar.gz
63da72119d2ff82095a4309bdbc92623067c0f8187be1ce56ba7ddbe9ff11960  guix-build-77e553ab6a0a/output/x86_64-apple-darwin/SHA256SUMS.part
935697e705a322fec78048fb1df34d5fd00aa9353afce508a387f7b7958eb7a9  guix-build-77e553ab6a0a/output/x86_64-apple-darwin/bitcoin-77e553ab6a0a-x86_64-apple-darwin-codesigning.tar.gz
a77aeff6f41b5d9b4278cdb1417b59520a6f09bef360063f3a69d3e213addf26  guix-build-77e553ab6a0a/output/x86_64-apple-darwin/bitcoin-77e553ab6a0a-x86_64-apple-darwin-unsigned.tar.gz
df59e2ac7f5676a7753ffba9571e2a503d177171d80e6d2260f595645ce367bb  guix-build-77e553ab6a0a/output/x86_64-apple-darwin/bitcoin-77e553ab6a0a-x86_64-apple-darwin-unsigned.zip
439ce4c9c108527d1c1f26f70b575e27d91321a41bad824bac5adff658bd8d29  guix-build-77e553ab6a0a/output/x86_64-linux-gnu/SHA256SUMS.part
011a4bd59cac89991163b0e04bfdb2f75ca1f590cd7880bf46f1ee6522c69351  guix-build-77e553ab6a0a/output/x86_64-linux-gnu/bitcoin-77e553ab6a0a-x86_64-linux-gnu-debug.tar.gz
7f1e7dbf479e939e6ffb1cceb7fd80e9bca8a186be8e6b7cf4d99d468eedc8a8  guix-build-77e553ab6a0a/output/x86_64-linux-gnu/bitcoin-77e553ab6a0a-x86_64-linux-gnu.tar.gz
aa9ceb2a5d639edf7a222810c2d4ffc5305cc274333153145606ff4ad77669bd  guix-build-77e553ab6a0a/output/x86_64-w64-mingw32/SHA256SUMS.part
f75a42a14138aa927f9483f0d5a4b2d160c9ee61bfc2715b8551f8eace90a563  guix-build-77e553ab6a0a/output/x86_64-w64-mingw32/bitcoin-77e553ab6a0a-win64-codesigning.tar.gz
9bf895bd1f16d7b8524213090e770dc7bdb11a1803f2358613d34aab6ba5435a  guix-build-77e553ab6a0a/output/x86_64-w64-mingw32/bitcoin-77e553ab6a0a-win64-debug.zip
05bff16e246264f47f29a0e1097ebcb666cbc34a07173dffff17061d1a2a2c08  guix-build-77e553ab6a0a/output/x86_64-w64-mingw32/bitcoin-77e553ab6a0a-win64-setup-unsigned.exe
bd29ba767147ccd87e1c43209e3a80b357aedbe9fb508d70e506704d72697fbe  guix-build-77e553ab6a0a/output/x86_64-w64-mingw32/bitcoin-77e553ab6a0a-win64-unsigned.zip
```
